File: backend/app/schemas/portfolio.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# --- Asset Snapshot ---

class AssetSnapshotResponse(BaseModel):
    id: uuid.UUID
    snapshot_date: date
    total_krw: float
    breakdown: dict
    created_at: datetime

    model_config = {"from_attributes": True}


class AssetTimelineResponse(BaseModel):
    snapshots: list[AssetSnapshotResponse]
    period: str
    start_date: date
    end_date: date


# --- Goal Asset ---

class GoalAssetCreate(BaseModel):
    target_amount: Decimal = Field(gt=0)
    target_date: date | None = None


class GoalAssetUpdate(BaseModel):
    target_amount: Decimal | None = Field(default=None, gt=0)
    target_date: date | None = None


class GoalAssetResponse(BaseModel):
    id: uuid.UUID
    target_amount: float
    target_date: date | None
    current_amount: float
    achievement_rate: float
    remaining_amount: float
    monthly_required: float | None
    estimated_date: date | None
    created_at: datetime
    updated_at: datetime


# --- Portfolio Target ---

class PortfolioTargetCreate(BaseModel):
    asset_type: str
    target_ratio: Decimal = Field(ge=0, le=1)


class PortfolioTargetBulkCreate(BaseModel):
    targets: list[PortfolioTargetCreate]


class PortfolioTargetResponse(BaseModel):
    id: uuid.UUID
    asset_type: str
    target_ratio: float
    current_ratio: float
    deviation: float
    created_at: datetime
    updated_at: datetime


class RebalancingAnalysisResponse(BaseModel):
    targets: list[PortfolioTargetResponse]
    total_deviation: float
    needs_rebalancing: bool
    threshold: float
    suggestions: list[dict]


# --- Rebalancing Alert ---

class RebalancingAlertResponse(BaseModel):
    id: uuid.UUID
    snapshot_date: date
    deviations: dict
    suggestion: dict
    threshold: float
    is_read: bool
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/shared/types/portfolio.ts
export interface AssetSnapshotResponse {
  id: string;
  snapshot_date: string;
  total_krw: number;
  breakdown: Record<string, number>;
  created_at: string;
}

export interface AssetTimelineResponse {
  snapshots: AssetSnapshotResponse[];
  period: string;
  start_date: string;
  end_date: string;
}

export interface GoalAssetCreate { target_amount: number; target_date?: string | null; }
export interface GoalAssetUpdate { target_amount?: number | null; target_date?: string | null; }

export interface GoalAssetResponse {
  id: string;
  target_amount: number;
  target_date: string | null;
  current_amount: number;
  achievement_rate: number;
  remaining_amount: number;
  monthly_required: number | null;
  estimated_date: string | null;
  created_at: string;
  updated_at: string;
}

export interface PortfolioTargetCreate { asset_type: string; target_ratio: number; }
export interface PortfolioTargetBulkCreate { targets: PortfolioTargetCreate[]; }

export interface PortfolioTargetResponse {
  id: string;
  asset_type: string;
  target_ratio: number;
  current_ratio: number;
  deviation: number;
  created_at: string;
  updated_at: string;
}

export interface RebalancingAnalysisResponse {
  targets: PortfolioTargetResponse[];
  total_deviation: number;
  needs_rebalancing: boolean;
  threshold: number;
  suggestions: Record<string, unknown>[];
}

export interface RebalancingAlertResponse {
  id: string;
  snapshot_date: string;
  deviations: Record<string, number>;
  suggestion: Record<string, unknown>;
  threshold: number;
  is_read: boolean;
  created_at: string;
}

// Backwards-compatible aliases (legacy names)
export type AssetTimeline = AssetTimelineResponse;
export type GoalAsset = GoalAssetResponse;
export type GoalAssetRequest = GoalAssetCreate;
export type PortfolioTarget = PortfolioTargetResponse;
export type PortfolioTargetRequest = PortfolioTargetCreate;
export type RebalancingAnalysis = RebalancingAnalysisResponse;
export type RebalancingAlert = RebalancingAlertResponse;
